### ISSUE TITLE
Interactive JupyTool and notebook (Jupyter Notebook) fails if output file is created

### ISSUE THREAD

**Issue by paulzierep** [galaxyproject/galaxy#16309]
**Describe the bug**
If an output is created (collection or normal), the notebook fails, the tool returns the output correctly, but the jupyter tool itself is red, blocking further work with the created files, even the executed notebook can be viewed, although it's shown as red. 
If the notebook is run without created output, it executes without error. 

After submitting a bug report, the Job Traceback shows:

```
Traceback (most recent call last):
  File "/opt/galaxy/server/lib/galaxy/jobs/runners/pulsar.py", line 676, in finish_job
    job_wrapper.finish(
  File "/opt/galaxy/server/lib/galaxy/jobs/__init__.py", line 1864, in finish
    self._finish_dataset(output_name, dataset, job, context, final_job_state, remote_metadata_directory)
  File "/opt/galaxy/server/lib/galaxy/jobs/__init__.py", line 1686, in _finish_dataset
    self.external_output_metadata.load_metadata(
  File "/opt/galaxy/server/lib/galaxy/metadata/__init__.py", line 301, in load_metadata
    import_model_store = store.imported_store_for_metadata(
  File "/opt/galaxy/server/lib/galaxy/model/store/__init__.py", line 2907, in imported_store_for_metadata
    import_model_store.perform_import()
  File "/opt/galaxy/server/lib/galaxy/model/store/__init__.py", line 383, in perform_import
    self._import_jobs(object_import_tracker, history)
  File "/opt/galaxy/server/lib/galaxy/model/store/__init__.py", line 1224, in _import_jobs
    imported_job.tool_id = job_attrs["tool_id"]
KeyError: 'tool_id'
```

It seems, that there is a problem with the collection of the metadata, which might explain, why the output is collected and the produced files are shown, even though the tool is show as failed.

Could not reproduce on https://usegalaxy.org/ since version 1.0.0 is not installed there, and 0.3 does not allow to store data in an output folder.

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 23.0 (usegalaxy.eu)
Commit: 

**To Reproduce**

# File output fail

1. Run interactive_tool_jupyter_notebook (https://usegalaxy.eu/root?tool_id=interactive_tool_jupyter_notebook) (version 1.0.0) with default values
2. Open the interactive tool
3. Open the launcher 
4. Open a terminal
5. Create a file `echo "A" > outputs/A.txt`
6. Open File / Shut Down
7. Look at created history

![Selection_021](https://github.com/galaxyproject/galaxy/assets/22075451/3814c428-edab-4d8a-a60f-d79055d6353c)

# Collection output fail

1. Run interactive_tool_jupyter_notebook (https://usegalaxy.eu/root?tool_id=interactive_tool_jupyter_notebook) (version 1.0.0) with default values
2. Open the interactive tool
3. Open the launcher 
4. Open a terminal
5. Create a file `echo "A" > outputs/collection/A.txt`
6. Create a file `echo "B" > outputs/collection/B.txt`
7. Open File / Shut Down
8. Look at created history

![Selection_021](https://github.com/galaxyproject/galaxy/assets/22075451/f4e5ff1c-32be-498b-ac90-7b79678b396c)
![Selection_022](https://github.com/galaxyproject/galaxy/assets/22075451/08294352-32cb-46f4-be01-dfa75bfd674b)

# Tool does not fail without output

1. Run interactive_tool_jupyter_notebook (https://usegalaxy.eu/root?tool_id=interactive_tool_jupyter_notebook) (version 1.0.0) with default values
2. Open the interactive tool
7. Open File / Shut Down
8. Look at created history

![Selection_023](https://github.com/galaxyproject/galaxy/assets/22075451/bed46f61-1bab-4e6b-a957-5d4c6e3e6545)

**Expected behavior**
The produces the correct output, but should not fail.

**Comment by bgruening** [comment]
This seems to happen after the outputs are collected and only if there are outputs that need to be collected like in the Jupytool.

@tuncK I assume that does not happen because you are not using the pulsar runner.

I have a second tool where this happens from @marie59 so it is not tool specific. But they are in both cases ITs with dynamic collections, ala:

```
        <collection name="outputs" type="list" label="ODV outputs">
            <discover_datasets pattern="__name_and_ext__" directory="./output" />
        </collection>
```

@mvdbeek do you have an idea here? Or can you install the latest JupyTool version on ORG so we can test? Thanks.

**Comment by paulzierep** [comment]
Still seems to be the case on EU as of now (cannot test on org since interactive_tool_jupyter_notebook is not instelled there): 
Stored files in collection folder:
```
Traceback (most recent call last):
  File "/opt/galaxy/server/lib/galaxy/jobs/runners/pulsar.py", line 685, in finish_job
    job_wrapper.finish(
  File "/opt/galaxy/server/lib/galaxy/jobs/__init__.py", line 1959, in finish
    self._finish_dataset(output_name, dataset, job, context, final_job_state, remote_metadata_directory)
  File "/opt/galaxy/server/lib/galaxy/jobs/__init__.py", line 1780, in _finish_dataset
    self.external_output_metadata.load_metadata(
  File "/opt/galaxy/server/lib/galaxy/metadata/__init__.py", line 301, in load_metadata
    import_model_store = store.imported_store_for_metadata(
  File "/opt/galaxy/server/lib/galaxy/model/store/__init__.py", line 2913, in imported_store_for_metadata
    import_model_store.perform_import()
  File "/opt/galaxy/server/lib/galaxy/model/store/__init__.py", line 387, in perform_import
    self._import_jobs(object_import_tracker, history)
  File "/opt/galaxy/server/lib/galaxy/model/store/__init__.py", line 1229, in _import_jobs
    imported_job.tool_id = job_attrs["tool_id"]
KeyError: 'tool_id'
```

**Comment by mvdbeek** [comment]
This is the old traceback.

**Comment by paulzierep** [comment]
@mvdbeek what do you mean ?

**Comment by mvdbeek** [comment]
I mean that whatever job handler is handling your job is not running off the latest 23.1 branch.

**Comment by paulzierep** [comment]
thanks for the clarification

**Comment by bgruening** [comment]
It is, pretty sure. But was the patch already merged forward to 23.1? I will update the branch, to be sure. But I think on Monday/Tuesday the patch was not in 23.1

**Comment by mvdbeek** [comment]
I think I merged it forward yesterday: https://github.com/galaxyproject/galaxy/commit/6d067f47b5fed4479332912c0009eea048f7a25a
